cql3/restrictions/statement_restrictions.cc

```diff
@@ -1003,5 +1006,725 @@ ::shared_ptr<single_column_restriction> single_column_restriction::LIKE::apply_t
     return r;
 }
 
+
+namespace {
+
+using cql3::selection::selection;
+
+/// Serialized values for all types of cells, plus selection (to find a column's index) and options (for
+/// subscript term's value).
+struct row_data_from_partition_slice {
+    const std::vector<bytes>& partition_key;
+    const std::vector<bytes>& clustering_key;
+    const std::vector<bytes_opt>& other_columns;
+    const selection& sel;
+};
+
+/// Data used to derive cell values from a mutation.
+struct row_data_from_mutation {
+    // Underscores avoid name clashes.
+    const partition_key& partition_key_;
+    const clustering_key_prefix& clustering_key_;
+    const row& other_columns;
+    const schema& schema_;
+    gc_clock::time_point now;
+};
+
+/// Everything needed to compute column values during restriction evaluation.
+struct column_value_eval_bag {
+    const query_options& options; // For evaluating subscript terms.
+    std::variant<row_data_from_partition_slice, row_data_from_mutation> row_data;
```

Comment:
Materialized views.  See https://github.com/scylladb/scylla/issues/6429#issuecomment-627392571 .

Reply:
Ho hum. This is pretty sad.

@psarna any ideas how to de-uglify it?

Reply:
One way is to convert the mutation form to data form, but it penalizes view maintenance (but speeds up normal queries).

Another way is to normalize them both to the same form (using bytes_view_opt).

Reply:
And, how is it handled now? I don't remember any dual-form code now.

Reply:
I luckily never had to touch "view filtering" before, so I don't have any clear ideas yet how to de-uglify it. I think the whole problem boils down to this piece from view.cc:
```cpp
bool matches_view_filter(const schema& base, const view_info& view, const partition_key& key, const clustering_row& update, gc_clock::time_point now) {
    return clustering_prefix_matches(base, view, key, update.key(), now)
            && boost::algorithm::all_of(
                view.select_statement().get_restrictions()->get_non_pk_restriction() | boost::adaptors::map_values,
                [&] (auto&& r) {
                    return r->is_satisfied_by(base, key, update.key(), update.cells(), cql3::query_options({ }), now);
                });
}
```

Reply:
> And, how is it handled now? I don't remember any dual-form code now.

Grep for is_satisfied_by, and you'll see dual versions everywhere.

Reply:
I see the sin was committed by 7be0f319d4f7da878efca7309053c158d6063afc, which adds a second way to access the restrictions machinery.

Reply:
I think the best way forward is to change this matches_view_filter to convert from mutations to values, and then we have a single type system for restrictions. This conversion will be expensive, but so will going through the visitor everywhere.

Reply:
> > And, how is it handled now? I don't remember any dual-form code now.
> 
> Grep for is_satisfied_by, and you'll see dual versions everywhere.

Yes. I think it is worse in `expression`, because the mv path limited it to a sub-tree (just single column restrictions, which is all that mv supports), but here, it proliferates everywhere.

Reply:
> This conversion will be expensive, but so will going through the visitor everywhere.

By "going through the visitor", do you mean "passing eval_bag"?  Because all the visitor code will still be there even after we unify the two is_satisfied_by entry points.  The only thing we'll dedupe is get_value_from_{mutation|partition}.

Reply:
Suppose we change matches_view_filter to convert the `row` that it has to the form that the regular cql3 machinery uses (which appears to be `const std::vector<bytes>& pk, const std::vector<bytes>& ck, const query::result_row_view& static_row, const query::result_row_view* row`, from result_set_builder::restrictions_builder, but maybe there are even other forms that are accepted at other points). Then the variant disappears, and there is only row_data.

Reply:
> Then the variant disappears, and there is only row_data.

Agreed, but all that changes is `eval_bag`->`row_data`, the rest of the code (visitors and all) stays exactly the same.  Only `get_value()` ends up simpler.

Reply:
And the runtime, which doesn't have to unpeel the bag each time.

Reply:
Should we do it in this PR or a follow-up?

Reply:
I'm okay with a follow-up. I'd prefer a pre-PR patch that does the simplification, but will accept a follow-up.